src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -358,23 +370,139 @@ def setClean(self, wflow):
+        It makes a query to LogDB in order to verify there are no any records for
+        the current workflow
+        :param wflow:       A MSRuleCleaner workflow representation
+        :return:            Bool
+        """
+        cleanStatus = False
+        logDBRecords = self.logDB.get(wflow['RequestName'])
+        self.logger.debug("logDBRecords: %s", pformat(logDBRecords))
+        if logDBRecords:
+            cleanStatus = True
+        return cleanStatus
+
+    def setLogDBClean(self, wflow):
+        """
+        A function to set the 'IsLogDBClean' flag based on the presence of any
+        records in LogDB for the current workflow.
+        :param  wflow:      A MSRuleCleaner workflow representation
+        :return wflow:
+        """
+        wflow['IsLogDBClean'] = self._checkLogDBClean(wflow)
+        return wflow
+
+    def findTargStatus(self, wflow):
+        """
+        Find the proper targeted archival status
+        :param  wflow:      A MSRuleCleaner workflow representation
+        :return wflow:
+        """
+        # Check the available status transitions before we decide the final status
+        targetStatusList = RequestStatus.REQUEST_STATE_TRANSITION.get(wflow['RequestStatus'], [])
+        targetStatusReg = re.compile(r'.*archived')
+        for status in targetStatusList:
+            if targetStatusReg.match(status):
+                wflow['TargetStatus'] = status
+        self.logger.debug("TargetStatus: %s", wflow['TargetStatus'])
+        return wflow
+
+    def _cleanLogDB(self, wflow):
+        """
+        A function to be used for cleaning all the records related to a workflow in logDB.
+        :param wflow:       A MSRuleCleaner workflow representation
+        :return:            Bool
+        """
+        cleanStatus = False
+        try:
+            self.logger.info("Deleting %s from LogDB WMStats...", wflow['RequestName'])
+            res = self.logDB.delete(wflow['RequestName'], agent=False)
+            if res == 'delete-error':
+                self.logger.error("Failed to delete logDB docs for wflow: %s", wflow['RequestName'])
+            cleanStatus = True
+        except Exception as ex:
+            msg = "General Exception while cleaning LogDB records for wflow: %s : %s"
+            self.logger.exception(msg, wflow['RequestName'], str(ex))
+        return cleanStatus
+
+    def _checkArchDelayExpired(self, wflow):
+        """
+        A function to check Archival Expiration Delay based on the inromation
+        returned by WMStatsServer regarding the time of the last request status transition
+        :param wflow:      MSRuleCleaner workflow representation
+        :return:           Bool
+        """
+        archDelayExpired = False
+        currentTime = int(time.time())
+        threshold = self.msConfig['archiveDelayHours'] * 3600
+        outputMask = ['RequestTransition']
+        inputConditon = {'RequestName': wflow['RequestName']}
```

Comment:
This will make one call per workflow, but we should actually make a single call per status, and simply go through the workflows.

Reply:
Ok, this one actually took me longer than I would like spend but anyway I think I managed to solve it in the best possible manner. I had to sacrifice memory footprint for run time complexity, but at the end I figured out that we can afford it.  @amaltaro please take a look and give me your opinion.

Reply:
This was remade yet again. To use `reqmgr` instead of `wmstatsServer` as @amaltaro was suggesting, and is actually a better looking piece now. Resolving this conversation now.